### PR TITLE
Monitor channels job

### PR DESCRIPTION
**Pull request by RodriFS**
The idea for this PR is to be able to check for and add ghost channels to the NodeGuard

### CODE REVIEW

File: src/Jobs/ChannelMonitorJob.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                    {
+                        Name = foundNode.Alias,
+                        PubKey = foundNode.PubKey,
+                    };
+                    var addNode = await _nodeRepository.AddAsync(node2);
+                    if (!addNode.Item1)
+                    {
+                        throw new Exception(addNode.Item2);
+                    }
+                }
```
Comment (daliclovr) [inline src/Jobs/ChannelMonitorJob.cs:96]: I think this logic of creating a new node if not found in the repository should be extracted, maybe creating a method in the repository if makes sense that actually creates a new one passing a lightning client? Or at least extracting it outside of this function if it doesn't makes sense in that scope, what do you think?

File: src/Jobs/ChannelMonitorJob.cs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        }
+
+        _logger.LogInformation("{JobName} ended", nameof(ChannelMonitorJob));
+    }
+
+    private async Task RecoverGhostChannels(Node source, Node destination, Channel? channel)
+    {
+        if (!channel.Initiator) return;
+
+        await using var context = await _dbContextFactory.CreateDbContextAsync();
```
Comment (daliclovr) [inline src/Jobs/ChannelMonitorJob.cs:124]: i would chose a better name such as `dbContext` 😄

File: src/Services/LightningService.cs [hunk 3]
```diff
@@ -635,6 +562,96 @@ await foreach (var response in openStatusUpdateStream.ResponseStream.ReadAllAsyn
...
+            if (channelUpdate.Item1 == false)
+            {
+                _logger.LogError(
+                    "Could not assign channel id to channel operation request: {RequestId} reason: {Reason}",
+                    channelOperationRequest.Id,
+                    channelUpdate.Item2);
+            }
+        }
+
+        public static async Task<Channel> CreateChannel(Node source, int destId, ChannelPoint channelPoint, long satsAmount, string? closeAddress = null)
```
Comment (daliclovr) [inline src/Services/LightningService.cs:620]: Is this the best place to put this logic? 🤔  just thinking

**Review by daliclovr — CHANGES_REQUESTED** [review]

File: src/Jobs/ChannelMonitorJob.cs [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        var data = context.JobDetail.JobDataMap;
+        var nodeId = data.GetInt("nodeId");
+        _logger.LogInformation("Starting {JobName} for node {nodeId}... ", nameof(ChannelMonitorJob), nodeId);
+        try
+        {
+            var node1 = await _nodeRepository.GetById(nodeId);
+
+            if (node1 == null)
+            {
+                _logger.LogInformation("The node: {@Node} is no longer ready to be supported quartz jobs", node1);
```
[the remark this answers is not in the document]
Comment (RodriFS) [reply src/Jobs/ChannelMonitorJob.cs:64]: I copy-pasted the information message, but it means that there is a job created for a node (maybe just recently) deleted from the NodeGuard. I should change the message.

File: src/Jobs/ChannelMonitorJob.cs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                    if (!addNode.Item1)
+                    {
+                        throw new Exception(addNode.Item2);
+                    }
+                }
+
+                try {
+                    // Recover Operations on channels
+                    await RecoverGhostChannels(node1, node2, channel);
+                    await RecoverChannelInConfirmationPendingStatus(node1);
```
[the remark this answers is not in the document]
Comment (RodriFS) [reply src/Jobs/ChannelMonitorJob.cs:101]: good catch, I'll create a try-catch inside the functions instead so they can move on if the previous one fails

File: src/Services/LightningService.cs (hunk 3 above)
Comment (RodriFS) [reply src/Services/LightningService.cs:620]: It was there already, I just moved it to it's own function. I agree that this method is insanely large, so we should start thinking about moving stuff

File: src/Pages/Channels.razor [hunk 6]
```diff
@@ -765,7 +765,8 @@
     private bool checkDisableCloseChannelButton(Channel channel)
     {
         ChannelOperationRequest? lastRequest = channel.ChannelOperationRequests.LastOrDefault();
-        if (lastRequest == null && channel.CreatedByNodeGuard == false) return false;
+        if (channel.CreatedByNodeGuard == false) return true;
+        if (lastRequest == null) return false;
```
Comment (RodriFS) [inline src/Pages/Channels.razor:769]: Changed this. But I'm not sure why we disable the button

File: src/Pages/Channels.razor [hunk 7]
```diff
@@ -717,7 +717,7 @@
 
     private bool OnWalletFilter(object itemValue, object searchValue)
     {
-        if (itemValue == null) return false;
+        if (itemValue == null) return true;
```
Comment (RodriFS) [inline src/Pages/Channels.razor:720]: for now ghost channels will appear with "Unknown" for which wallet opened the channel

File: src/Jobs/ChannelMonitorJob.cs [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        {
+            var node1 = await _nodeRepository.GetById(nodeId);
+
+            if (node1 == null)
+            {
+                _logger.LogInformation("The node {NodeId} was set up for monitoring but the node doesn't exist anymore", nodeId);
+                return;
+            }
+
+            var client = LightningService.CreateLightningClient(node1.Endpoint);
```
Comment (Jossec101) [inline src/Jobs/ChannelMonitorJob.cs:68]: Hmm.. creating a gRPC client all the time can lead to resources exhaustion, we should add them via DI

File: src/Pages/Channels.razor (hunk 7 above)
Comment (Jossec101) [reply src/Pages/Channels.razor:720]: But, the " ghost" channels were created by NG, we have the info. I don't get this.

File: src/Pages/Channels.razor (hunk 6 above)
Comment (Jossec101) [reply src/Pages/Channels.razor:769]: Me neither, worth checking, good point, we shall be able to close channels created from outside.

File: src/Helpers/Constants.cs [hunk 9]
```diff
@@ -55,6 +55,7 @@ public class Constants
 
     // Crons & Jobs
     public static readonly string MONITOR_WITHDRAWALS_CRON = "10 0/5 * * * ?";
+    public static readonly string MONITOR_CHANNELS_CRON = "0 0 0 * * ?";
```
Comment (Jossec101) [inline src/Helpers/Constants.cs:58]: That's all nights at 00:00AM, shoudn't this be more periodic? like every hour?

File: src/Pages/Channels.razor (hunk 7 above)
Comment (RodriFS) [reply src/Pages/Channels.razor:720]: If we have a request in the `ChannelOperationRequest` table, then we know the wallet that originated the funding Tx. But if the channel was originated by lncli for example, or the node had previous channels before it began to be managed by the NodeGuard, then we won't have the info about which wallet created it. But we still would want to manage those channels, right?

File: src/Jobs/ChannelMonitorJob.cs (hunk 8 above)
Comment (RodriFS) [reply src/Jobs/ChannelMonitorJob.cs:68]: Will investigate

File: src/Pages/Channels.razor (hunk 7 above)
Comment (Jossec101) [reply src/Pages/Channels.razor:720]: Agree +1

File: src/Jobs/ChannelMonitorJob.cs (hunk 8 above)
Comment (RodriFS) [reply src/Jobs/ChannelMonitorJob.cs]: I created a service for reusing clients in jobs.
Other parts of the application (like the lightning service) create the lightning client inside static functions, so we will have to refactor this in order to reuse the client there.

File: src/Helpers/Constants.cs (hunk 9 above)
Comment (Jossec101) [reply src/Helpers/Constants.cs:62]: Hmm, I would do it more often

Comment (RodriFS) [reply src/Helpers/Constants.cs:62]: moving to every hour

**Comment by RodriFS** [comment]
Added unit testing

File: src/Jobs/ChannelMonitorJob.cs (hunk 8 above)
Comment (Jossec101) [reply src/Jobs/ChannelMonitorJob.cs]: Why not this? https://learn.microsoft.com/en-us/aspnet/core/grpc/clientfactory?view=aspnetcore-7.0

**Review by Jossec101 — CHANGES_REQUESTED** [review]
Lets use gRPC client factory by MSFT to have a good way to use gRPC channels efficiently:

https://learn.microsoft.com/en-us/aspnet/core/grpc/clientfactory?view=aspnetcore-7.0

**Review by daliclovr — DISMISSED** [review]

**Review by Jossec101 — APPROVED** [review]

**Pull request merged**